File: cpp/src/ch08/game.cc
#include <memory>
#include <iostream>
#include "game.h"

using std::cout;
using std::endl;

double play_game(AIFunction actions_wb[2],
                 StateVersion state_versions[2],
                 bool should_print)
{
    std::array<std::unique_ptr<ConnectFourState>, 2> states = {
        get_state(state_versions[0]),
        get_state(state_versions[1])};
    // auto state = ConnectFourState();

    if (should_print)
        cout << states[0]->to_string() << endl;

    int player = 0;
    while (!states[0]->is_done())
    {
        int action;
        if (player == 0)
            action = actions_wb[player](states[0]);
        else
            action = actions_wb[player](states[1]);

        states[0]->advance(action);
        states[1]->advance(action);

        if (should_print)
            cout << states[0]->to_string() << endl;

        player ^= 1; // change player
    }
    return states[0]->white_score();
}

double many_games(AIFunction actions_wb[2],
                  StateVersion state_versions[2],
                  int num_games,
                  int print_every)
{
    double total = 0;
    for (int i = 0; i < num_games; i++)
    {
        total += play_game(actions_wb, state_versions, false);

        if (print_every > 0 && i % print_every == 0)
            cout << "i " << i << " w " << total / (i + 1) << endl;
    }
    return total / num_games;
}

double games_black_and_white(AIFunction actions_wb[2],
                             StateVersion state_versions[2],
                             int num_games,
                             int print_every)
{
    AIFunction actions_bw[2] = {actions_wb[1], actions_wb[0]};
    double total = 0;

    cout << "white" << endl;
    total += many_games(actions_wb, state_versions, num_games, print_every);

    cout << "black" << endl;
    total += 1 - many_games(actions_bw, state_versions, num_games, print_every);

    return total / 2;
}

File: python/src/thunder_book/ch08/game.py
import logging
from typing import Type

from thunder_book.ch08.maze_state import (
    ActionFunc,
    MazeParams,
    MazeState,
)
from thunder_book.util import printerr

logger = logging.getLogger(__name__)


def play_game(
    params: MazeParams,
    actions: tuple[ActionFunc, ActionFunc],
    state_types: tuple[Type[MazeState], Type[MazeState]],
    should_print: bool = False,
) -> float:
    # state = ConnectFourState(params)
    states = [state_types[0](params), state_types[1](params)]

    if should_print:
        logger.debug(states[0].to_string())

    player = 0
    while not states[0].is_done():
        action = actions[player](states[player])
        for state in states:
            state.advance(action)
        player ^= 1
        if should_print:
            logger.debug(states[0].to_string())
            printerr(states[0].to_string())

    return state.white_score()


def play_many(
    params: MazeParams,
    actions: tuple[ActionFunc, ActionFunc],
    state_types: tuple[Type[MazeState], Type[MazeState]],
    num_games: int,
    print_every: int = 10,
) -> float:
    total_score = 0.0
    for i in range(num_games):
        total_score += play_game(params, actions, state_types)

        if print_every > 0 and (i + 1) % print_every == 0:
            logger.debug(f"{i+1} {total_score / (i+1):.2f}")
    return total_score / num_games


def play_black_and_white(
    params: MazeParams,
    actions: tuple[ActionFunc, ActionFunc],
    state_types: tuple[Type[MazeState], Type[MazeState]],
    num_games: int,
    print_every: int = 10,
) -> float:
    total = play_many(params, actions, state_types, num_games, print_every)
    actions_bw = (actions[1], actions[0])
    total += 1 - play_many(params, actions_bw, state_types, num_games, print_every)
    return total / 2
